runtime/lua/vim/fcnotify.lua

```diff
@@ -0,0 +1,224 @@
+-- Create a file watcher, each watcher is identified by the name of the file that it
+-- watches.
+
+local uv = vim.loop
+local Watcher = {}
+local w = {}
+local WatcherList = {}
+local commence = nil
+
+--- Callback for the check handle, checks if there are pending notifications
+--- for any watcher, and handles them as per the value of the `fcnotify`
+--- option.
+local function check_notifications()
+  for _, watcher in pairs(WatcherList) do
+    local option = vim.api.nvim_buf_get_option(watcher.bufnr, 'filechangenotify')
+    if watcher.pending_notifs and watcher.paused == false and option ~= 'off' then
+      if uv.fs_stat(watcher.ffname) ~= nil then
+        vim.api.nvim_command('checktime '..watcher.bufnr)
+        watcher.pending_notifs = false
+      else
+        error("ERR: File "..watcher.fname.." removed")
+      end
+    end
+  end
+end
+
+--- Checks if a buffer should have a watcher attached to it.
+--- Ignores all the buffers that aren't listed, or have a buf id
+--- less than 0.
+--- Ignores the [NO NAME] buffer.
+---
+--@param fname: The validity of this buffer will be checked.
+local function valid_buf(fname)
+  if fname == '' then
+    return false
+  end
+
+  local bufnr = vim.api.nvim_call_function('bufnr', {fname})
+
+  if bufnr < 0 then
+    return false
+  end
+
+  local buflisted = vim.api.nvim_buf_get_option(bufnr, 'buflisted')
+  local buftype = vim.api.nvim_buf_get_option(bufnr, 'buftype')
+
+  if not buflisted or buftype == 'nofile' or buftype == 'quickfix' then
+    return false
+  end
+  return true
+end
+
+local check_handle = uv.new_check()
+check_handle:start(vim.schedule_wrap(check_notifications))
+
+--- Creates and initializes a new watcher object with the given filename.
+---
+--@param fname: (required, string) The path that the watcher should watch.
+function Watcher:new(fname)
+  vim.validate{fname = {fname, 'string', false}}
+  -- get full path name for the file
+  local ffname = vim.api.nvim_call_function('fnamemodify', {fname, ':p'})
+  w = {bufnr = vim.api.nvim_call_function('bufnr', {fname}),
+       fname = fname, ffname = ffname, handle = nil,
+       paused = false, pending_notifs = false}
+  setmetatable(w, self)
+  self.__index = self
+  return w
+end
+
+local function start_poll(fname)
+  WatcherList[fname].handle = uv.new_fs_poll()
+  WatcherList[fname].handle:start(WatcherList[fname].ffname, 200, function(...)
+    WatcherList[fname]:poll_change(...)
+  end)
+end
+
+--- Starts the watcher
+---
+--@param self: (required, table) The watcher object on which start was called.
+local function start_watch(fname)
+  WatcherList[fname].handle = uv.new_fs_event()
+  WatcherList[fname].handle:start(WatcherList[fname].ffname, {}, function(...)
+    WatcherList[fname]:watch_change(...)
+  end)
+end
+
+function Watcher:start()
+  if self.handle and not self.handle:is_closing() then
+    self.handle:close()
+  end
+
+  commence(self.fname)
+end
+
+function Watcher.init_poll()
```

Comment:
Can you change these names to something like `_set_warching_mechanism`? I don't care the exact name, but it's not _exactly_ `init` something, since it looks like you can change it any time you want.

Reply:
i discussed this with @teto and decided that I'll add polling handle after this PR is merged.
(PS: The naming really got me on this one. I had never been so confused. When I will add this later on, I'll probably spread things out between different files. There were so many starts in this one that I had to resolve to `commence` for a function name xD)